stuff/cron/build_problem_rec_model.py

```diff
@@ -0,0 +1,333 @@
+                AND r.status = "ready"
+                AND r.verdict = "AC";
+            """, dbconn)
+        logging.info('Found %d runs', len(runs))
+
+        # Split dataset into test/train
+        users = pd.Series(runs.identity_id.unique())
+        train_users = users.sample(frac=self.config.train_fraction,
+                                   random_state=self.rng_seed)
+        test_users = users.drop(train_users.index)
+        logging.info('Training users: %d', len(train_users))
+        logging.info('Testing users: %d', len(test_users))
+
+        return runs, train_users, test_users
+
+    def save(self, output_path: str) -> None:
+        '''Save a model to an Sqlite3 db in `output_path`.
+
+        The DB contains a single table `ProblemRecommendations` with 3 columns:
+        - solved_problem_id (int): The ID of the problem the user just solved.
+        - recommended_problem_id (int): One recommendation.
+        - score (float): A relative weight for this recommendation vs
+                         others.
+
+        The DB also has an index on `solved` for efficient lookups.
+        '''
+        with sqlite3.connect(output_path) as conn:
+            try:
+                cur = conn.cursor()
+                cur.executescript(
+                    '''CREATE TABLE ProblemRecommendations (
+                        solved_problem_id INTEGER,
+                        recommended_problem_id INTEGER,
+                        score REAL
+                    );
+                    CREATE INDEX Recs_index ON ProblemRecommendations(solved);
+                    ''')
+                for solved_problem_id, recs in self.model:
+                    for recommended_problem_id, score in recs:
+                        cur.execute('''
+                            INSERT INTO
+                                ProblemRecommendations
+                            VALUES
+                                (?, ?, ?)
+                            ''',
+                                    (solved_problem_id, recommended_problem_id,
+                                     score))
+            finally:
+                conn.commit()
+
+    def generate_weighted_pairs(self) -> pd.DataFrame:
+        '''Assumes runs are sorted by submission time'''
+        tuples = []
+        for _, problems in self.train_ac.groupby('identity_id'):
+            num_problems = len(problems)
+            # TODO: Figure out how to ask Pandas nicely for this.
+            for i in range(num_problems):
+                source = problems.iloc[i]['problem_id']
+                weight = 1.0
+                for j in range(1, self.config.num_followups + 1):
+                    if i + j >= num_problems:
+                        break
+                    current = problems.iloc[i + j]['problem_id']
+                    tuples.append((source, current, weight))
+                    weight *= self.config.followup_decay
+
+        output = pd.DataFrame(tuples)
```

Comment:
```suggestion
        return pd.DataFrame(tuples, columns=('source', 'target', 'weight'))
```

así puedes hacerlo un one-liner.

Reply:
Oh!